File: engine/src/ecs/transform.c
/*
    This file is a part of yoyoengine. (https://github.com/zoogies/yoyoengine)
    Copyright (C) 2023-2025  Ryan Zmuda

    Licensed under the MIT license. See LICENSE file in the project root for details.
*/

#include <stddef.h>
#include <stdlib.h>

#include <yoyoengine/ecs/ecs.h>
#include <yoyoengine/ecs/transform.h>

void ye_add_transform_component(struct ye_entity *entity, int x,int y){
    entity->transform = malloc(sizeof(struct ye_component_transform));
    // entity->transform->active = true; transform doesnt need active
    entity->transform->x = x;
    entity->transform->y = y;

    // add this entity to the transform component list
    ye_entity_list_add(&transform_list_head, entity);

    // log that we added a transform and to what ID
    // ye_logf(debug, "Added transform to entity %d\n", entity->id);
}

void ye_remove_transform_component(struct ye_entity *entity){
    free(entity->transform);
    entity->transform = NULL;

    // remove the entity from the transform component list
    ye_entity_list_remove(&transform_list_head, entity);
}

File: engine/src/lua_runtime/ecs/transform.lua
--[[
    This file is a part of yoyoengine. (https://github.com/zoogies/yoyoengine)
    Copyright (C) 2023-2025  Ryan Zmuda

    Licensed under the MIT license. See LICENSE file in the project root for details.
]]

---@class Transform
---@field x number The x position
---@field y number The y position
Transform = {
    -- no **real** fields.
    -- This exists purely for intellisense
}

-- define the transform metatable
Transform_mt = {
    __index = function(self, key) 
        local parent_ptr = rawget(self, "parent_ptr")

        -- if not ValidateEntity(parent) then
        --     log("error", "Transform field accessed on nil/null entity\n")
        --     return nil
        -- end

        local x = ye_lua_transform_get_position_x(parent_ptr)
        local y = ye_lua_transform_get_position_y(parent_ptr)

        if key == "x" then
            return x
        elseif key == "y" then
            return y
        else
            log("error", "Transform field accessed with invalid key\n")
            return nil
        end
    end,

    __newindex = function(self, key, value)
        local parent_ptr = rawget(self, "parent_ptr")

        -- if not ValidateEntity(parent) then
        --     log("error", "Transform field accessed on nil/null entity\n")
        --     return
        -- end

        if key == "x" then
            ye_lua_transform_set_position_x(parent_ptr, value)
        elseif key == "y" then
            ye_lua_transform_set_position_y(parent_ptr, value)
        else
            log("error", "Transform field accessed with invalid key\n")
            return
        end
    end,
}

---**Create a new Transform component.**
---
--- You must pass either x AND y or neither.
---
---@param x? number The x position of the Transform
---@param y? number The y position of the Transform
function Entity:AddTransformComponent(x, y) end -- fake prototype for intellisense
function AddTransformComponent(self, x, y)
    if x and y then
        Entity:addComponent(self, ye_lua_create_transform, x, y)
    else
        Entity:addComponent(self, ye_lua_create_transform, 0, 0)
    end
end

---**Remove the Transform component from the entity.**
function Entity:RemoveTransformComponent() end -- fake prototype for intellisense
function RemoveTransformComponent(self)
    ye_lua_remove_component(rawget(self, "_c_entity"), TRANSFORM_COMPONENT)
end